src/Microsoft.OData.Core/Json/BufferingJsonReader.cs

```diff
@@ -999,6 +1350,589 @@ private void RemoveFirstNodeInBuffer()
+                    bool wasBuffering = this.isBuffering;
+                    BufferedNode storedPosition = null;
+                    if (this.isBuffering)
+                    {
+                        storedPosition = this.currentBufferedNode;
+                    }
+                    else
+                    {
+                        await this.StartBufferingAsync()
+                            .ConfigureAwait(false);
+                    }
+
+                    await this.ProcessObjectValueAsync()
+                        .ConfigureAwait(false);
+
+                    // Reset the reader state to non-buffering or to the previously
+                    // backed up position in the buffer.
+                    if (wasBuffering)
+                    {
+                        this.currentBufferedNode = storedPosition;
+                    }
+                    else
+                    {
+                        this.StopBuffering();
+                    }
+                }
+
+                return result;
+            }
+            finally
+            {
+                this.parsingInStreamError = false;
+            }
+        }
+
+        /// <summary>
+        /// Try to read an error structure from the stream. Return null if no error structure can be read.
+        /// </summary>
+        /// <param name="error">An <see cref="ODataError"/> instance that was read from the reader or null if none could be read.</param>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains a tuple comprising of:
+        /// 1). A value of true if an <see cref="ODataError"/> instance that was read; otherwise false.
+        /// 2). An <see cref="ODataError"/> instance that was read from the reader or null if none could be read.</returns>
+        private async Task<Tuple<bool, ODataError>> TryReadInStreamErrorPropertyValueAsync()
+        {
+            Debug.Assert(this.parsingInStreamError, "this.parsingInStreamError");
+            this.AssertBuffering();
+            this.AssertAsynchronous();
+
+            // We expect a start-object node here
+            if (this.currentBufferedNode.NodeType != JsonNodeType.StartObject)
+            {
+                return Tuple.Create(false, (ODataError)null);
+            }
+
+            // Read the start-object node
+            await this.ReadInternalAsync()
+                .ConfigureAwait(false);
+
+            ODataError error = new ODataError();
+
+            // We expect one of the supported properties for the value (or end-object)
+            ODataJsonLightReaderUtils.ErrorPropertyBitMask propertiesFoundBitmask = ODataJsonLightReaderUtils.ErrorPropertyBitMask.None;
+            while (this.currentBufferedNode.NodeType == JsonNodeType.Property)
+            {
+                Tuple<bool, string> readErrorStringPropertyResult;

```

Comment:
Maybe this variable should be defined where it's used instead of at the top of the switch case? So we don't have to keep it in mind until we need it (when reading the code).